docs/pages/blog/react-19-update.md

```diff
@@ -0,0 +1,142 @@
+    Component.displayName = render.displayName ?? render.name;
+    return Component as React.ForwardRefExoticComponent<P>;
+  }
+  return React.forwardRef(
+    render as React.ForwardRefRenderFunction<T, React.PropsWithoutRef<P>>,
+  );
+};
+```
+
+The shim provides two key benefits:
+
+1. Type safety - TypeScript will warn if props are spread incorrectly
+2. Forward compatibility - Components using the shim will work correctly in all supported React versions
+
+The code update was:
+
+```tsx
+// Before
+const GridRoot = forwardRef((props, ref) => {
+  const state = useGridState();
+  return <div ref={ref} {...props} {...state} />;
+});
+
+// After
+const GridRoot = forwardRefShim((props, ref) => {
+  const state = useGridState();
+  return <div {...props} {...state} ref={ref} />;
+});
+```
+
+## Phase 2: Moving to React 19
+
+After ensuring compatibility, we started working on migrating the codebase to React 19. This involved:
+
+1. Updating all package dependencies to their React 19 compatible versions
+2. Migrating test utilities to work with React 19
+3. Ensuring all components work with the new React 19 features
+4. Updating CI to run tests with React 18
+5. Type references with `RefObject` for React 19 and with `MutableRefObject` for earlier versions.
+
+The biggest change in this phase was around the `useRef()` hook [update](https://react.dev/blog/2024/04/25/react-19-upgrade-guide#useref-requires-argument) and the fact that `apiRef` in the DataGrid component had to be updated from `MutableRefObject` to `RefObject` for React 19 only, in order to avoid type errors for users (and their CI) that did not migrate yet.
+
+As part of the migration we also moved to NextJS 15 for our documentation website.
+
+### Our own `RefObject`
+
+To have different types for `apiRef` in the DataGrid component for different React versions, we created our own `RefObject` type.
+
+We used the fact that `useRef()` requires a param for React 19 to have `RefObject` evaluated to `MutableRefObject` for React < 19 and to `RefObject` otherwise.
+
+```ts
+// in React 19 useRef requires a parameter, so `() => any` will not match anymore
+export type RefObject<T> = typeof React.useRef extends () => any
+  ? React.MutableRefObject<T>
+  : React.RefObject<T>;
+```
+
+## Conclusion
+
+The migration to React 19 was a significant undertaking that required careful planning and execution. By breaking it down into two phases we were able to quickly provide React 19 compatibility for our users while we worked on our own migration.
+
+The migration has also helped us identify areas where we can further improve our components' type safety and performance.
+
+It is worth noting that a big chunk of our own migration was done by the MUI Core team as they added support for React 19 to both `@mui/material` v5 and v6. Additionally, they provided the necessary updates to the internal tools that both of our repositories use for building and testing our components.
```

Comment:
I think we should avoid using internal team names like "MUI Core" in this context because it's probably just confusing to most readers. If we want to shout out the Core team here I think we should just say "the maintainers of Material UI," and we should probably go further to name MUI X in this context to help reinforce the distinction. Here's how I'd write this:

```suggestion
Though this project was spearheaded by the MUI X team, we owe a special thanks to our colleagues who maintain Material UI for their massive help in adding React 19 support to both v5 and v6 of `@mui/material`.
They also provided the necessary updates to the internal tools that both of our repositories use for building and testing our components.
```